File: code/chap07/datasource_elasticsearch_reader.py
from __future__ import print_function 
import sys 
from pyspark.sql import SparkSession 
from pyspark.sql import Row
import json

#-----------------------------------------------------
# Read From ElasticSearch and write to an RDD
# Input: ElasticSearch Hostname
#------------------------------------------------------
# Input Parameters:
#    ElasticSearch Hostname
#-------------------------------------------------------
# @author Mahmoud Parsian
#-------------------------------------------------------
def main():

    if len(sys.argv) != 2:  
        print("Usage: datasource_elasticsearch_reader.py <es-hostname>", file=sys.stderr)
        exit(-1)

    # read name of ElasticSearch Hostname
    es_hostname = sys.argv[1]
    print("es_hostname : ", es_hostname)

    # create an instance of SparkSession
    spark = SparkSession.builder.getOrCreate()   
    
    # Read from ElasticSearch and Write to RDD

    # Now that we have some data in Elasticsearch
    # (by running datasource_elasticsearch_writer.py), 
    # we can read it back in using the elasticsearc-hadoop 
    # connector.
    #
    es_read_conf = {
        # specify the node that we are sending data to 
        # (this should be the master node)
        "es.nodes" : es_hostname,
        # specify the port in case it is not the default port
        "es.port" : '9200',
        # specify a resource in the form 'index/doc-type'
        "es.resource" : 'testindex/testdoc'
    }
    
    # Now we can generate an RDD from this Elasticsearch data:
    es_rdd = spark.sparkContext.newAPIHadoopRDD(
        inputFormatClass="org.elasticsearch.hadoop.mr.EsInputFormat",
        keyClass="org.apache.hadoop.io.NullWritable",
        valueClass="org.elasticsearch.hadoop.mr.LinkedMapWritable",
        conf=es_read_conf
    )
    #
    print("rs_rdd : ",  es_rdd)
    print("es_rdd.count() : ",  es_rdd.count())
    print("es_rdd.collect() \n: ", es_rdd.collect())
    
    # done!
    spark.stop()
#end-def

if __name__ == '__main__':
    main()

File: code/chap07/scala/src/main/scala/org/data/algorithms/spark/ch07/DatasourceElasticsearchReader.scala
package org.data.algorithms.spark.ch07

import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.io.NullWritable
import org.apache.spark.sql.SparkSession
import org.elasticsearch.hadoop.mr.LinkedMapWritable

/**
 *-----------------------------------------------------
 * Read From ElasticSearch and write to an RDD
 * Input: ElasticSearch Hostname
 *------------------------------------------------------
 * Input Parameters:
 *    ElasticSearch Hostname
 *-------------------------------------------------------
 *
 * @author Biman Mandal
 *-------------------------------------------------------
 */
object DatasourceElasticsearchReader {

  def main(args: Array[String]): Unit = {
    if(args.length != 2) {
      System.err.println("Pass Elasticsearch Server and Port in Command Line Argument")
      System.exit(-1)
    }

    // read name of ElasticSearch Hostname
    val esHostname = args(0)
    println("esHostname : " + esHostname)

    val esPort = args(1)
    println("esPort : " + esPort)

    // create an instance of SparkSession
    val spark = SparkSession.builder()
      .master("local[*]")
      .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
      .config("spark.kryo.classesToRegister", "org.apache.hadoop.io.Text")
      .getOrCreate()

    /**
     * Read from ElasticSearch and Write to RDD
     *
     * Now that we have some data in Elasticsearch
     * (by running DatasourceElasticsearchWriter),
     * we can read it back in using the elasticsearc-hadoop
     * connector.
     *
     */
    val esReadConf = new Configuration()
    // specify the node that we are sending data to
    // (this should be the master node)
    esReadConf.set("es.nodes", esHostname)
    // specify the port in case it is not the default port
    esReadConf.set("es.port", esPort)
    // Comment it out in case your elastic search is not running in local docker
    esReadConf.set("es.nodes.wan.only", "true")
    // specify a resource index name
    esReadConf.set("es.resource", "testindex")

    // Now we can generate an RDD from this Elasticsearch data:
    val esRdd = spark.sparkContext.newAPIHadoopRDD(
      esReadConf,
      classOf[org.elasticsearch.hadoop.mr.EsInputFormat[NullWritable,LinkedMapWritable]],
      classOf[org.apache.hadoop.io.NullWritable],
      classOf[org.elasticsearch.hadoop.mr.LinkedMapWritable]
    )
    println("esRdd : " +  esRdd)
    println("esRdd.count() : " +  esRdd.count())
    println("esRdd.collect() \n: " + esRdd.collect().mkString("Array(", ", ", ")"))

    // done!
    spark.stop()


  }
}
